internal/extsvc/types.go

```
@@ -488,6 +488,53 @@ func WebhookURL(kind string, externalServiceID int64, cfg any, externalURL strin
 	return fmt.Sprintf("%s/.api/%s?%s=%d%s", externalURL, path, IDParam, externalServiceID, extra), nil
 }
 
+func ExtractEncryptableBaseURL(ctx context.Context, kind string, config *EncryptableConfig) (CodeHostBaseURL, error) {
+	parsed, err := ParseEncryptableConfig(ctx, kind, config)
+	if err != nil {
+		return CodeHostBaseURL{}, errors.Wrap(err, "loading service configuration")
+	}
+
+	url, err := extractURL(parsed, kind)
+	if err != nil {
+		return CodeHostBaseURL{}, err
+	}
+	if url == "" {
+		return CodeHostBaseURL{}, errors.New("service does not have a URL")
+	}
+
+	baseUrl, err := NewCodeHostBaseURL(url)
+	return baseUrl, err
+}
+
+func extractURL(parsed any, kind string) (string, error) {
+	switch c := parsed.(type) {
+	case *schema.GitHubConnection:
+		return c.Url, nil
+	case *schema.BitbucketServerConnection:
+		return c.Url, nil
+	case *schema.BitbucketCloudConnection:
+		return c.Url, nil
+	case *schema.PagureConnection:
+		return c.Url, nil
+	case *schema.PerforceConnection:
+		return "", nil
```

as per documentation of `ExternalService.BaseURL` method, these cases should return an error, right?
Should we fix the doc or fix the code here and return an error?

---

it does! further up we check whether the URL is `""` and if so return an error.

---

oh, this function is called from `ExtractEncryptableBaseURL` which handles errors, sorry, my bad :)